lib/trento/hosts/host.ex

```diff
@@ -172,19 +172,24 @@ defmodule Trento.Hosts.Host do
           installation_source: installation_source
         }
       ) do
-    %HostRegistered{
-      host_id: host_id,
-      hostname: hostname,
-      ip_addresses: ip_addresses,
-      agent_version: agent_version,
-      cpu_count: cpu_count,
-      total_memory_mb: total_memory_mb,
-      socket_count: socket_count,
-      os_version: os_version,
-      installation_source: installation_source,
-      fully_qualified_domain_name: fully_qualified_domain_name,
-      heartbeat: :unknown
-    }
+    host
+    |> Multi.new()
+    |> Multi.execute(fn _ ->
+      %HostRegistered{
+        host_id: host_id,
+        hostname: hostname,
+        ip_addresses: ip_addresses,
+        agent_version: agent_version,
+        cpu_count: cpu_count,
+        total_memory_mb: total_memory_mb,
+        socket_count: socket_count,
+        os_version: os_version,
+        installation_source: installation_source,
+        fully_qualified_domain_name: fully_qualified_domain_name,
+        heartbeat: :unknown
+      }
+    end)
+    |> Multi.execute(&maybe_emit_software_updates_discovery_requested_event/1)
```

Comment:
What if we group all of these maybes at once:

```
# args are, host_id, new_fqdn, old_fqdn
def maybe_emit_software_updates_discovery_events(host_id, nil, nil),do: nil

def maybe_emit_software_updates_discovery_events(host_id, nil, _old_fqdn) do
  %SoftwareUpdatesDiscoveryCleared{
    host_id: host_id
  }
end

def maybe_emit_software_updates_discovery_events(host_id, new_fqdn, nil) do
  %SoftwareUpdatesDiscoveryRequested{
    host_id: host_id,
    fully_qualified_domain_name: new_fqdn
  }
end

def maybe_emit_software_updates_discovery_events(host_id, fqdn, fqdn) do: nil

def maybe_emit_software_updates_discovery_events(host_id, new_fqdn, _old_fqdn) do
  %SoftwareUpdatesDiscoveryRequested{
    host_id: host_id,
    fully_qualified_domain_name: new_fqdn
  }
end
```

No `is_nil` functions, just standard easy pattern matching.

Reply:
This proposal always takes into account a transition of the fqdn, but sometimes we just care about the end result in the aggregate state other times we need to check the transition, as we discussed in another comment.

This is why the multi and the functions as they have been implemented now. 

What could we do differently?

Reply:
As far as I see using the `multi` is not necessary.
If the software update events must be emitted depends simply on the old and new `fqdn` values, and we don't need a state update to get that, they are already there, in the initial state and event.

For example, if you substitute 1:1 the functions i wrote there by the ones you used, all the tests are green. Not a big deal, but by default, it is better to not use `when` guards if you can solve things with simple pattern matching, and using less functions looks ideal (my opinion).

It is true that not using the `multi` is a bit inconvenient on how you need to use this functions. As you cannot send a `nil` event in a list of events, you need to something like this:

```
def execute(
        %Host{host_id: nil},
        %RegisterHost{
          host_id: host_id,
          ...
        }
      ) do
    [
      %HostRegistered{
        host_id: host_id,
         ...
      }
    ] ++ maybe_emit_software_updates_discovery_events(host_id, fully_qualified_domain_name, nil)
  end
```

and later on change slightly the functions I wrote to return lists:
```
  def maybe_emit_software_updates_discovery_events(_host_id, nil, nil), do: []

  def maybe_emit_software_updates_discovery_events(host_id, nil, _old_fqdn) do
    [
      %SoftwareUpdatesDiscoveryCleared{
        host_id: host_id
      }
    ]
  end
```
and so on. For me this looks perfectly fine.
Tests are green in this case as well.

If it was up to me, I would choose this option, as you don't use a multi that you don't really need, and from my perspective, it is cleaner and precise (by the way, using multi is not harmful, simply, we don't need it in my opinion).

Now, this is my opinion, your solution works (as does mine), and as this is your PR, I leave the decision up to you.

PD: i base my opinion seeing the tests coming in this PR are all green